x-pack/legacy/plugins/advanced_ui_actions/public/np_ready/public/custom_time_range_badge.tsx

```diff
@@ -0,0 +1,92 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import React from 'react';
+import { prettyDuration, commonDurationRanges } from '@elastic/eui';
+
+import { TimeRange } from '../../../../../../../src/plugins/data/public';
+import {
+  Action,
+  IEmbeddable,
+  ActionContext,
+  IncompatibleActionError,
+  Embeddable,
+  EmbeddableInput,
+} from '../../../../../../../src/legacy/core_plugins/embeddable_api/public/np_ready/public';
+
+import { CustomizeTimeRangeModal } from './customize_time_range_modal';
+import { doesInheritTimeRange } from './does_inherit_time_range';
+import { OpenModal, CommonlyUsedRange } from './types';
+
+const CUSTOM_TIME_RANGE_BADGE = 'CUSTOM_TIME_RANGE_BADGE';
+
+export interface TimeRangeInput extends EmbeddableInput {
+  timeRange: TimeRange;
+}
+
+function hasTimeRange(
+  embeddable: IEmbeddable | Embeddable<TimeRangeInput>
+): embeddable is Embeddable<TimeRangeInput> {
+  return (embeddable as Embeddable<TimeRangeInput>).getInput().timeRange !== undefined;
+}
+
+export class CustomTimeRangeBadge extends Action {
+  public readonly type = CUSTOM_TIME_RANGE_BADGE;
+  private openModal: OpenModal;
+  private dateFormat: string;
+  private commonlyUsedRanges: CommonlyUsedRange[];
+
+  constructor({
+    openModal,
+    dateFormat,
+    commonlyUsedRanges,
+  }: {
+    openModal: OpenModal;
+    dateFormat: string;
+    commonlyUsedRanges: CommonlyUsedRange[];
+  }) {
+    super(CUSTOM_TIME_RANGE_BADGE);
+    this.order = 7;
+    this.openModal = openModal;
+    this.dateFormat = dateFormat;
+    this.commonlyUsedRanges = commonlyUsedRanges;
+  }
+
+  public getDisplayName({ embeddable }: ActionContext<Embeddable<TimeRangeInput>>) {
+    return prettyDuration(
+      embeddable.getInput().timeRange.from,
+      embeddable.getInput().timeRange.to,
+      commonDurationRanges,
+      this.dateFormat
+    );
+  }
+
+  public getIconType() {
+    return 'calendar';
+  }
+
+  public async isCompatible({ embeddable }: ActionContext) {
+    return Boolean(embeddable && hasTimeRange(embeddable) && !doesInheritTimeRange(embeddable));
+  }
+
+  public execute({ embeddable }: ActionContext) {
+    if (!this.isCompatible({ embeddable })) {
```

Comment:
```suggestion
    if (!(await this.isCompatible({ embeddable }))) {
```

This branch will never execute, because `this.isCompatible()` returns a `Promise` which is truthy.